server/queue/queue-rabbitmq/src/main/java/org/apache/james/queue/rabbitmq/managable/cassandra/EnqueuedMailsDao.java

```diff
@@ -0,0 +1,391 @@
+                .map(bucketAndSlice -> browse(queueName, bucketAndSlice.getBucketId(), bucketAndSlice.getSliceStartInstant())))
+            .thenApply(stream -> stream.flatMap(Function.identity()));
+    }
+
+    private CompletableFuture<Stream<ManageableMailQueue.MailQueueItemView>> browse(
+        MailQueueName queueName, Integer bucketId, Instant sliceStartInstant) {
+
+        CompletableFuture<Stream<CassandraManageableMailQueueItem>> enqueuedItemsFuture = executor.execute(
+            selectFrom.bind()
+                .setString(QUEUE_NAME, queueName.asString())
+                .setTimestamp(TIME_RANGE_START, Date.from(sliceStartInstant))
+                .setInt(BUCKET_ID, bucketId))
+            .thenApply(resultSet -> cassandraUtils.convertToStream(resultSet)
+                .map(EnqueuedMailsDaoUtil::toMailQueueItemView));
+
+        return enqueuedItemsFuture
+            .thenCompose(this::filterNonDeletedItem);
+    }
+
+    private CompletableFuture<Stream<ManageableMailQueue.MailQueueItemView>> filterNonDeletedItem(
+        Stream<CassandraManageableMailQueueItem> manageableMailQueueItemStream) {
+
+        return CompletableFutureUtil.allOf(
+            manageableMailQueueItemStream
+                .map(deletedMailsDao::filterNotDeleted))
+            .thenApply(this::filterPresentItem);
+    }
+
+    private Stream<ManageableMailQueue.MailQueueItemView> filterPresentItem(
+        Stream<Optional<CassandraManageableMailQueueItem>> afterCheckingDeletionStream) {
+
+        return afterCheckingDeletionStream
+            .filter(Optional::isPresent)
+            .map(Optional::get)
+            .map(CassandraManageableMailQueueItem::getMailQueueItemView);
+    }
+
+    private Stream<BucketAndSlice> bucketAndSliceStream(Date timeRangeStartFrom) {
+        Stream<Instant> sliceStartStream = sliceStartStream(timeRangeStartFrom);
+        Stream<Integer> bucketStream = IntStream.range(0, manageableQueueConfiguration.getBucketCount()).boxed();
+
+        return bucketStream
+            .flatMap(bucketId -> sliceStartStream
+                .map(sliceStartInstant -> BucketAndSlice.of(bucketId, sliceStartInstant)));
+    }
+
+    private Stream<Instant> sliceStartStream(Date timeRangeStartFrom) {
+        List<Instant> sliceWindows = new ArrayList<>();
+
+        int sliceWindowInterval = manageableQueueConfiguration.getSliceWindowSize();
+
+        Instant now = Instant.now();
+        Instant nextSlice = sliceStart(timeRangeStartFrom, sliceWindowInterval);
+
+        while (nextSlice.isBefore(now)) {
+            sliceWindows.add(Instant.ofEpochSecond(nextSlice.getEpochSecond()));
+            nextSlice = nextSlice.plusSeconds(sliceWindowInterval);
+        }
+
+        return sliceWindows.stream();
+    }
+
+    private Instant sliceStart(Date timeRangeStartFrom, int sliceWindowInterval) {
+        long sliceIdStart = timeRangeStartFrom.toInstant().getEpochSecond() / sliceWindowInterval;
```

Comment:
Variable naming makes a great difference for understanding what you are writing.

Here you are computing the number of slice since UNIX time reference.

Thus `sliceIdStart` should be named `sliceCount` (as it is not a time)